R/meta_utils.R

```diff
@@ -0,0 +1,90 @@
+#' Extract the result metadata of a Tplyr table
+#'
+#' Given a row_id value and a result column, this function will return the
+#' tplyr_meta object associated with that 'cell'.
+#'
+#' If a Tplyr table is built with the `metadata=TRUE` option specified, then
+#' metadata is assembled behind the scenes to provide traceability on each
+#' result cell derived. The functions `get_meta_result()` and
+#' `get_meta_subset()` allow you to access that metadata by using an ID provided
+#' in the row_id column and the column name of the result you'd like to access.
+#' The purpose is of the row_id variable instead of a simple row index is to
+#' provide a sort resistant reference of the originating column, so the output
+#' Tplyr table can be sorted in any order but the metadata are still easily
+#' accessible.
+#'
+#' The `tplyr_meta` object provided a list with two elements - names and
+#' filters. The metadata contain every column from the target data.frame of the
+#' Tplyr table that factored into the specified result cell, and the filters
+#' contains all the necessary filters to subset to data summarized to create the
+#' specified result cell. `get_meta_subset()` additionally provides a parameter to
+#' specify any additional columns you would like to include in the returned
+#' subset data frame.
+#'
+#'
+#' @param t A built Tplyr table
+#' @param row_id The row_id value of the desired cell, provided as a character
+#'   string
+#' @param column The result column of interest, provided as a character string
+#'
+#' @return A tplyr_meta object for `get_meta_result()`, and a tibble for
+#'   `get_meta_subset()`
+#' @md
+#'
+#' @family Metadata Extraction Tools
+#' @rdname meta_extraction
+#' @export
+#'
+#' @examples
+#' t <- tplyr_table(mtcars, cyl) %>%
+#'   add_layer(
+#'     group_desc(hp)
+#'   )
+#'
+#' dat <- t %>% build(metadata = TRUE)
+#'
+#' get_meta_result(t, 'd1_1', 'var1_4')
+#'
+#' get_meta_subset(t, 'd1_1', 'var1_4', add_cols = dplyr::vars(carb))
+get_meta_result <- function(t, row_id, column) {
+  m <- t$metadata
+
+  if (!inherits(row_id, 'character') || !(row_id %in% m$row_id)) {
+    stop('Invalid row_id selected. row_id must be provided as a string present in built Tplyr table.',
+         call.=FALSE)
+  }
+
+  if (!inherits(column, 'character') || !(column %in% names(m))) {
+    stop(paste0('column must provided as a character string and a valid result ',
+         'column present in the built Tplyr dataframe'), call.=FALSE)
+  }
+
+  # Pull out the cell of interest
+  res <- m[[which(m$row_id == row_id), column]][[1]]
+
+  if (!inherits(res, 'tplyr_meta')) {
+    stop('Specified column must be a result column', call.=FALSE)
+  }
+
+  res
+}
+
+#' @param add_cols  Additional columns to include in subset data.frame output
+#'
+#' @family Metadata Extraction Tools
+#' @rdname meta_extraction
+#' @export
+get_meta_subset <- function(t, row_id, column, add_cols = vars(USUBJID)) {
```

Comment:
Instead of using `vars(USUBJID)` would it be better to use the `distinct` variable that's being used in the table?

Reply:
So it's only count layers that we'd ever supply a distinct variable. Any other table types the distinct variable serves no purpose, so I wanted the default to push back USUBJID because you're pretty much always going to want the subject number included in the data returned. 

If you don't like the explicitness of it, I can remove USUBJID and have users just add it explicitly. Should distinct_by be a variable generally captured for count layers though in the metadata? It would go in the `names` element and just not have a filter associated with it.